The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17711588882/job/50330996406) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50330996406) [(plain)](https://github.com/rust-lang/rust/commit/7d19a6d84d3e30b9c8e5fd751be34cb6f3c09fd7/checks/50330996406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] nix test:false 1.655
   Compiling rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
warning: ignoring -C extra-filename flag due to -o flag

error[E0425]: cannot find value `twostep` in this scope
   --> compiler/rustc_mir_transform/src/dest_prop.rs:555:49
    |
555 |                         values.insert(relevant, twostep);
    |                                                 ^^^^^^^ not found in this scope

error[E0425]: cannot find value `twostep` in this scope
   --> compiler/rustc_mir_transform/src/dest_prop.rs:558:49
    |
558 |                         values.insert(relevant, twostep);
    |                                                 ^^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
[RUSTC-TIMING] rustc_mir_transform test:false 0.625
warning: `rustc_mir_transform` (lib) generated 1 warning
```

</details>

--- a/compiler/rustc_mir_transform/src/dest_prop.rs
+++ b/compiler/rustc_mir_transform/src/dest_prop.rs
@@ -596,8 +596,14 @@ fn save_as_intervals<'tcx>(
             // as behaving so by default.
             // We make an exception for simple assignments `_a.stuff = {copy|move} _b.stuff`,
             // as marking `_b` live here would prevent unification.
-            let is_simple_assignment =
-                matches!(stmt.kind, StatementKind::Assign(box (_, Rvalue::Use(_))));
+            let is_simple_assignment = match stmt.kind {
+                StatementKind::Assign(box (
+                    lhs,
+                    Rvalue::CopyForDeref(rhs)
+                    | Rvalue::Use(Operand::Copy(rhs) | Operand::Move(rhs)),
+                )) => lhs.projection == rhs.projection,
+                _ => false,
+            };
             VisitPlacesWith(|place: Place<'tcx>, ctxt| {
                 if let Some(relevant) = relevant.shrink[place.local] {
                     match DefUse::for_place(place, ctxt) {

--- a/compiler/rustc_mir_transform/src/dest_prop.rs
+++ b/compiler/rustc_mir_transform/src/dest_prop.rs
@@ -590,8 +590,12 @@ fn save_as_intervals<'tcx>(
             twostep = TwoStepIndex::from_u32(twostep.as_u32() + 1);
             debug_assert_eq!(twostep, two_step_loc(loc, Effect::After));
             append_at(&mut values, &state, twostep);
-            // Ensure we have a non-zero live range even for dead stores. This is done by marking
-            // all the written-to locals as live in the second half of the statement.
+            // Like terminators, ensure we have a non-zero live range even for dead stores.
+            // Some rvalues interleave reads and writes, for instance `Rvalue::Aggregate`, see
+            // https://github.com/rust-lang/rust/issues/146383. By precaution, treat statements
+            // as behaving so by default.
+            // We make an exception for simple assignments `_a.stuff = {copy|move} _b.stuff`,
+            // as marking `_b` live here would prevent unification.
             let is_simple_assignment =
                 matches!(stmt.kind, StatementKind::Assign(box (_, Rvalue::Use(_))));
             VisitPlacesWith(|place: Place<'tcx>, ctxt| {

